### PR TITLE
PR: `Link.regex/0` `Link.find/1` and `Link.find_replace_compact/1` issue #4

### PR DESCRIPTION
**Pull request by nelsonic**
As the title suggests, this PR creates the functions we need for finding and compacting `URLs` in our `MVP` ref: https://github.com/dwyl/mvp/issues/141

### CODE REVIEW

**Comment by nelsonic** [comment]
@LuchoTurtle please review + merge when you can. Thanks. 👌

**Comment by nelsonic** [comment]
@LuchoTurtle on reading my own code I spotted a potential `Bug`: https://github.com/dwyl/link/issues/4#issuecomment-1659534045 🐞 
Gonna take a quick look at it this morning and then will re-assign back to you for review. 👌

**Comment by nelsonic** [comment]
Try it: https://amemo.fly.dev/
```markdown
This is my link:
https://github.com/dwyl/mvp/issues/141#issuecomment-1657954420
it gets compacted to save screen space.
But [markdown links](https://github.com/dwyl/mvp/issues/141#issuecomment-1657954420) 
and ![images](https://i.pinimg.com/1200x/5d/fe/59/5dfe590aa252b50f8e5b35a129f617d5.jpg) are unmodified.
Try it! 😊
```

For: https://github.com/dwyl/mvp/issues/141#issuecomment-1661489434

**Review by LuchoTurtle — APPROVED** [review]
Awesome work! Works great ⭐

**Comment by LuchoTurtle** [comment]
Thanks @nelsonic , works wonders!

I don't know if you want to support all of Github's content but two instances don't look exactly like what happens on Github.

Screenshot 2023-08-02 at 13 41 16


this is a PR - https://github.com/dwyl/link/pull/5

this is an issue - https://github.com/dwyl/flutter-wysiwyg-editor-tutorial/issues/1#issuecomment-1646888350

this is a link to a comment - https://github.com/dwyl/flutter-wysiwyg-editor-tutorial/issues/1#issuecomment-1646888350

this is a link to a review comment - https://github.com/dwyl/link/pull/5#pullrequestreview-1558913764

this is a repo - https://github.com/dwyl/link

Do you want me to merge this nonetheless? I think these are perhaps not needed for this PR but since you've gone through the trouble of implementing them, I'm wondering if it makes sense to treat these edge cases.

**Comment by nelsonic** [comment]
@LuchoTurtle Good edge cases. Let's add tests for them and write the code.

```sh
https://github.com/dwyl/link/pull/5 -> dwyl/link/pull/5

https://github.com/dwyl/flutter-wysiwyg-editor-tutorial/issues/1#issuecomment-1646888350 -> dwyl/flutter-wysiwyg-editor-tutorial#1

https://github.com/dwyl/flutter-wysiwyg-editor-tutorial/issues/1#issuecomment-1646888350 -> dwyl/flutter-wysiwyg-editor-tutorial#1

https://github.com/dwyl/link/pull/5#pullrequestreview-1558913764 -> dwyl/link/pull/5

https://github.com/dwyl/link -> dwyl/link
```

**Comment by nelsonic** [comment]
If we input these links:
```md
https://github.com/dwyl/link/pull/5  
https://github.com/dwyl/flutter-wysiwyg-editor-tutorial/issues/1#issuecomment-1646888350  
https://github.com/dwyl/link/pull/5#pullrequestreview-1558913764  
https://github.com/dwyl/link
```
Into the demo app: https://amemo.fly.dev we see that the Pull Request and Review Comment links aren't working:
 
image

I will add tests and code for these. 👨‍💻 ⏳

File: lib/link.ex [hunk 1]
```diff
@@ -61,16 +67,151 @@ defmodule Link do
...
+  (?:\?[^\n\r]*)?              # Optional query string ?q=
+  (?:#[^\n\r]*)?               # Optional fragment e.g: #comment
+
+  ## Examples
+
+      iex> Regex.run(Link.regex(), "dwyl.com") |> List.flatten |> Enum.filter(& &1 != "") |> List.first
+      "dwyl.com"
+  """
+  def regex do
+    ~r|[(http(s)?):\/\/(www\.)?a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,14}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)|
```
Comment (nelsonic) [inline lib/link.ex:144]: If anyone has a _better_ `regex` for detecting URLs please share! 🙏

**Comment by nelsonic** [comment]
@LuchoTurtle I've added the `tests` and `code` to make `PR` links work #7 
Test on the demo version: https://amemo.fly.dev/

image

```md
https://github.com/dwyl/link spacing text  
https://github.com/dwyl/link/pull/5 more text    
https://github.com/dwyl/link/pull/5#pullrequestreview-1558913764  
more random text and a markdown link
[link](https://github.com/dwyl/link) 
```

or with the list of links you shared above:

```md
https://github.com/dwyl/link
https://github.com/dwyl/link/pull/5  
https://github.com/dwyl/flutter-wysiwyg-editor-tutorial/issues/1#issuecomment-1646888350  
https://github.com/dwyl/link/pull/5#pullrequestreview-1558913764  
```

image

Please review & merge when you can. 🙏

**Review by LuchoTurtle — CHANGES_REQUESTED** [review]
Thanks for the changes 😍 .

I was testing the compacted Github links you've fixed and they seem to be working fine (although when there's a "review comment" or "regular comment", it doesn't show *exactly* like Github's with the `(comment)` string, but it's fine).

However, I've noticed a funny behaviour. Check the following video. It seems that when the string is empty, Github's compaction doesn't seem to work.


https://github.com/dwyl/link/assets/17494745/3f4ea904-39f6-4f33-8a10-c3c1c0b5c8b2

Is this desirable?
I'm going to `request changes` but I can try to get it fixed, if you want to.

**Comment by nelsonic** [comment]
@LuchoTurtle thanks for this valuable real-world QA-ing. I've opened an issue: https://github.com/dwyl/link/issues/8 🆕

**Comment by nelsonic** [comment]
BTW: removing the word "review" or "comment" from the hypertext is 100% desirable; they add no value in link text. 😉 ✂️

**Comment by nelsonic** [comment]
@LuchoTurtle please see: https://github.com/dwyl/link/issues/8#issuecomment-1666415465 🙏

**Comment by nelsonic** [comment]
https://github.com/dwyl/link/blob/d342f8fc83a67077b483f10bee906894e2578c28/lib/link.ex#L207-L208

File: lib/link.ex [hunk 2]
```diff
@@ -42,35 +49,182 @@ defmodule Link do
...
+
+  @doc """
+  `find/1` finds all instances of a URL in a block of text.
+
+  ## Examples
+
+      iex> Link.find("Text with links http://goo.gl/3co4ae and https://git.io/top & www.dwyl.com etc.")
+      ["http://goo.gl/3co4ae", "https://git.io/top", "www.dwyl.com"]
+
+      # Find a link without any other text or whitespace #8
```
Comment (nelsonic) [inline lib/link.ex:171]: Added this doctest for #8 
Passes without modifying any `code`. ✅

File: lib/link.ex [hunk 3]
```diff
@@ -42,35 +49,182 @@ defmodule Link do
...
+      iex> md = "# Hello World! https://github.com/dwyl/mvp/issues/141#issuecomment-1657954420 and https://mvp.fly.dev/ "
+      iex> Link.find_replace_compact(md)
+      "# Hello World! [dwyl/mvp#141](https://github.com/dwyl/mvp/issues/141#issuecomment-1657954420) and [mvp.fly.dev](https://mvp.fly.dev/) "
+
+      # Does not attempt to compact an existing markdown [link](https://github.com/dwyl/link) or ![image](https://imgur.com/gallery/odNLFdO)
+      iex> md = "existing markdown [link](https://github.com/dwyl/link) or ![image](https://imgur.com/gallery/odNLFdO)"
+      iex> Link.find_replace_compact(md)
+      "existing markdown [link](https://github.com/dwyl/link) or ![image](https://imgur.com/gallery/odNLFdO)"
+
+      iex> Link.find_replace_compact("https://github.com/dwyl/link/pull/5#pullrequestreview-1558913764")
```
Comment (nelsonic) [inline lib/link.ex:209]: Added this doctest for #8 ... passes without modifying any `code`. 🤷‍♂️

**Comment by nelsonic** [comment]
@LuchoTurtle PR links Working flawlessly on: https://amemo.fly.dev/
pr-links-still-work

File: lib/link.ex (hunk 3 above)
Comment (nelsonic) [reply lib/link.ex:209]: It's working. https://github.com/dwyl/link/pull/5#issuecomment-1666943043 ✅

File: lib/link.ex (hunk 2 above)
Comment (nelsonic) [reply lib/link.ex:171]: PR links are working. https://github.com/dwyl/link/pull/5#issuecomment-1666943043 ✅

**Comment by nelsonic** [comment]
@LuchoTurtle should be reviewable/mergeable now (when you get back from holidays). 👌

**Review by LuchoTurtle — APPROVED** [review]
lgtm!
Thanks for the changes! I've tried a few GH possibilities and it seems to work tremendously well! 🥳 

image

**Comment by nelsonic** [comment]
@LuchoTurtle Thanks for reviewing + merge. 🙏 
Follow-up issue created: https://github.com/dwyl/link/issues/9 🙃
